haystack/preview/components/embedders/openai_document_embedder.py

```diff
@@ -0,0 +1,157 @@
+        """
+        texts_to_embed = []
+        for doc in documents:
+            meta_values_to_embed = [
+                str(doc.metadata[key])
+                for key in self.metadata_fields_to_embed
+                if key in doc.metadata and doc.metadata[key]
+            ]
+
+            text_to_embed = (
+                self.prefix + self.embedding_separator.join(meta_values_to_embed + [doc.text or ""]) + self.suffix
+            )
+
+            # copied from OpenAI embedding_utils (https://github.com/openai/openai-python/blob/main/openai/embeddings_utils.py)
+            # replace newlines, which can negatively affect performance.
+            text_to_embed = text_to_embed.replace("\n", " ")
+            texts_to_embed.append(text_to_embed)
+        return texts_to_embed
+
+    def _embed_batch(self, texts_to_embed: List[str], batch_size: int) -> Tuple[List[str], Dict[str, Any]]:
+        """
+        Embed a list of texts in batches.
+        """
+
+        all_embeddings = []
+        metadata = {}
+        for i in tqdm(
+            range(0, len(texts_to_embed), batch_size), disable=not self.progress_bar, desc="Calculating embeddings"
+        ):
+            batch = texts_to_embed[i : i + batch_size]
+            response = openai.Embedding.create(model=self.model_name, input=batch)
+            embeddings = [el["embedding"] for el in response.data]
+            all_embeddings.extend(embeddings)
+
+            if "model" not in metadata:
+                metadata["model"] = response.model
+            if "usage" not in metadata:
+                metadata["usage"] = dict(response.usage.items())
+            else:
+                metadata["usage"]["prompt_tokens"] += response.usage.prompt_tokens
+                metadata["usage"]["total_tokens"] += response.usage.total_tokens
+
+        return all_embeddings, metadata
+
+    @component.output_types(documents=List[Document], metadata=Dict[str, Any])
+    def run(self, documents: List[Document]):
+        """
+        Embed a list of Documents.
+        The embedding of each Document is stored in the `embedding` field of the Document.
+        """
+        if not isinstance(documents, list) or not isinstance(documents[0], Document):
+            raise TypeError(
+                "OpenAIDocumentEmbedder expects a list of Documents as input."
+                "In case you want to embed a string, please use the OpenAITextEmbedder."
+            )
+
+        openai.api_key = self.api_key
+        if self.organization is not None:
+            openai.organization = self.organization
+
+        texts_to_embed = self._prepare_texts_to_embed(documents=documents)
+
+        embeddings, metadata = self._embed_batch(texts_to_embed=texts_to_embed, batch_size=self.batch_size)
+
+        documents_with_embeddings = []
+        for doc, emb in zip(documents, embeddings):
+            doc_as_dict = doc.to_dict()
```

Comment:
This is a downside of the immutability constraint we put on Documents. There is nothing we can do about it here.

Reply:
Adding an `add_embedding` method to the Document class
```
def add_embedding(self, embedding):
        object.__setattr__(self, 'embedding', embedding)
```
would probably work but is too hacky and against the `@dataclass(frozen=True)` concept. Do you know whether this was discussed earlier?

Reply:
I hit this wall, in fact :smiley: 
It has not been discussed before...
(Looping in @ZanSara to keep track of this)

Reply:
Hey! I've always intended this concept of frozen documents to be a test and see if we ever have issues with that. We're totally still in time to change this concept, but time is running out: we must be quite confident whether we can or want to roll with the frozen Document class before the 2.0 release.

The reasoning behind this choice is that the ID and the eventual embedding represent the content of a document. So IDs, embeddings and content should be difficult to change without changing the other. However, nowadays I see this "frozen" feature as mostly an obstacle too, because it's easy to make the mistake of updating the embedding without changing the content, and re-creating the class every time just slows us the component.

Imho if we've got this far is a good sign, but I'm open to un-freezing them if we believe is better. Keep in mind also that un-freezing them is an easier decision than freezing them back.

One more thing: if we un-freeze Document, then I think we should consider:

1. Tie more strongly the ID to the fields that generated it (so IDs don't go out of sync accidentally). This may mean making ID a property instead of a field?

2. introducing a way to track which model generated a given embedding. Maybe an Embedding class with two fields (`embedding` and `embedder_model`)? Another field in Document (`embedder_model`)? 
 
Just food for thought.

Reply:
https://github.com/deepset-ai/haystack/issues/5899